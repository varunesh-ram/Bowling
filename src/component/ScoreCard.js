import React from 'react';
import { Constants } from '../constants/Constants';

export default class ScoreCard extends React.Component {
    header = () => {
        const headerTitles = Constants.SCORECARD_TITLES.map((title) => (
            <th key={title} colSpan="6">
                {title}
            </th>
        ));
        return headerTitles;
    };

    pinsDownOnRolls = () => {
        const rolls = [];
        let i = 0;
        for (let frame = 0; frame < 10; frame++) {
            const colSpanValue = frame > 8 ? "2" : "3";
            if (frame === 9) {
                this.generateRollsForTenthFrame(rolls, frame, colSpanValue, i);
            }
            else {
                this.generateRollsForFirstNineFrames(rolls, frame, colSpanValue, i);
            }
            if (this.isStrike(i)) {
                i++;
            } else {
                i += 2;
            }
        }
        return rolls;
    };

    footer = () => {
        let bottomLine = [];
        for (let i = 0; i < 11; i++) {
            bottomLine.push(
                <td key={"frame" + i} colSpan="6">
                </td>
            );
        }
        return bottomLine;
    };

    isStrike(i) {
        return this.props.rolls[i] === 10;
    }

    generateRollsForFirstNineFrames(rolls, frame, colSpanValue, i) {
        rolls.push(
            <td key={(2 * frame)} id={"r" + (2 * frame)} colSpan={colSpanValue}>
                {this.props.rolls.length > i ? this.isStrike(i) ? "" : this.props.rolls[i] : ""}
            </td>
        );
        rolls.push(
            <td key={(2 * frame) + 1} id={"r" + ((2 * frame) + 1)} colSpan={colSpanValue}>
                {this.isStrike(i) ? "X" : this.props.rolls.length > i + 1 ? this.isSpar(i) ? "/" : this.props.rolls[i + 1] : ""}
            </td>);
    }

    isSpar(i) {
        return this.props.rolls[i] + this.props.rolls[i + 1] === 10;
    }

    generateRollsForTenthFrame(rolls, frame, colSpanValue, i) {
        rolls.push(
            <td key={(2 * frame)} id={"r" + (2 * frame)} colSpan={colSpanValue}>
                {this.props.rolls.length > i ? this.isStrike(i) ? "X" : this.props.rolls[i] : ""}
            </td>
        );
        rolls.push(
            <td key={(2 * frame) + 1} id={"r" + ((2 * frame) + 1)} colSpan={colSpanValue}>
                {this.props.rolls.length > i + 1 ? this.isSpar(i) ? "/" : this.isStrike(i+1) ? "X" : this.props.rolls[i + 1] : ""}
            </td>);
        rolls.push(
            <td key={20} id={"r" + (20)} colSpan={colSpanValue}>
                {this.props.rolls.length > i + 2 ? this.props.rolls[i + 2] === 10 ? "X" : this.props.rolls[i + 2] : ""}
            </td>);
    }

    render() {
        return (
            <div className="Container">
                <table id="table" className="Scorecard" cellPadding="1" cellSpacing="0">
                    <tbody>
                        <tr>{this.header()}</tr>
                        <tr>
                            {this.pinsDownOnRolls()}
                            <td id="total-score" colSpan="6">
                                {this.props.score}
                            </td>
                        </tr>
                        <tr>{this.footer()}</tr>
                    </tbody>
                </table>
            </div>
        );
    }
}