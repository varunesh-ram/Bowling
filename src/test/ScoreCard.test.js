import React from "react";
import { shallow } from "enzyme";
import ScoreCard from "../component/ScoreCard";

describe("Scorecard component", () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<ScoreCard rolls="" score=""u />);
    });

    it("should render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe("Render ScoreCard Table", () => {

        it("should render table with three rows", () => {
            expect(wrapper.find("tr").length).toEqual(3);
        });

        it("should render table header with 11 columns with heading", () => {
            expect(wrapper.find("tr").at(0).find("th").length).toEqual(11);
            expect(wrapper.find("th").at(0).text()).toEqual("Frame 1");
            expect(wrapper.find("th").at(1).text()).toEqual("Frame 2");
            expect(wrapper.find("th").at(2).text()).toEqual("Frame 3");
            expect(wrapper.find("th").at(3).text()).toEqual("Frame 4");
            expect(wrapper.find("th").at(4).text()).toEqual("Frame 5");
            expect(wrapper.find("th").at(5).text()).toEqual("Frame 6");
            expect(wrapper.find("th").at(6).text()).toEqual("Frame 7");
            expect(wrapper.find("th").at(7).text()).toEqual("Frame 8");
            expect(wrapper.find("th").at(8).text()).toEqual("Frame 9");
            expect(wrapper.find("th").at(9).text()).toEqual("Frame 10");
            expect(wrapper.find("th").at(10).text()).toEqual("Total Score");
        });

        it("should render table content with 22 cells (21 for rolls and 1 for total score) empty", () => {
            expect(wrapper.find("tr").at(1).find("td").length).toEqual(22);
            wrapper.find("tr").at(1).find("td").forEach((cell) => {
                expect(cell.text()).toEqual("");
            });
        });

        it("should render table footer with 11 cells empty", () => {
            expect(wrapper.find("tr").at(2).find("td").length).toEqual(11);
            wrapper.find("tr").at(2).find("td").forEach((cell) => {
                expect(cell.text()).toEqual("");
            });
        });
    });


});