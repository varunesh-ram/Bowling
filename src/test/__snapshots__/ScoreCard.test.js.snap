// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scorecard component should render correctly 1`] = `undefined`;
